### PR TITLE
Current main

### PR DESCRIPTION
**Pull request by andrew2003**

### CODE REVIEW

File: .github/workflows/rust-ci.yml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    steps:
+      # Get code from repo
+      - uses: actions/checkout@v4
+        # Set up Rust tool chain
+      - uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: nightly
+          components: clippy
+          override: true
```
Comment (zk-steve) [inline .github/workflows/rust-ci.yml:23]: .

File: Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  "migrate",
+  "postgres",
+] }
+tokio = { version = "1.1.1", features = ["full"] }
+tracing = { version = "0.1.40", features = ["log"] }
+tracing-bunyan-formatter = { version = "0.3.9" }
+tracing-opentelemetry = { version = "0.25.0" }
+tracing-subscriber = { version = "0.3.17", features = ["env-filter"] }
+warp = "0.3"
+
```
Comment (zk-steve) [inline Cargo.toml:33]: ?

File: src/controllers/company.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
+use crate::models::company::{Company, CompanyId, NewCompany};
+use crate::models::pagination::{Pagination, PaginationMethods};
+use crate::models::store_trait::StoreMethods;
+use crate::service::jwt::Claims;
+use crate::utils::convert_to_json::{Data, PayloadNoData, PayloadWithData};
+use std::collections::HashMap;
+use std::sync::Arc;
+use tracing::instrument;
+use warp::http::StatusCode;
```
Comment (zk-steve) [inline src/controllers/company.rs:9]: imports

File: src/controllers/company.rs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            StatusCode::BAD_REQUEST,
+        ));
+    }
+
+    let new_company = NewCompany {
+        email: new_company.email,
+        name: new_company.name,
+        address: new_company.address,
+        description: new_company.description,
+    };
```
Comment (zk-steve) [inline src/controllers/company.rs:39]: .

File: src/controllers/company.rs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            let payload = PayloadWithData {
+                message: "Success".to_string(),
+                data: Data::Company(res),
+            };
+            Ok(warp::reply::with_status(
+                warp::reply::json(&payload),
+                StatusCode::OK,
+            ))
+        }
+        Err(e) => Err(warp::reject::custom(e)),
```
Comment (zk-steve) [inline src/controllers/company.rs:146]: .

File: src/controllers/company.rs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+// returns a success response with status code 200 if the company is deleted successfully.
+#[instrument(level = "info", skip(store))]
+pub async fn delete_company(
+    store: Arc<dyn StoreMethods + Send + Sync>,
+    claims: Claims,
+    company: Company,
+) -> Result<impl warp::Reply, warp::Rejection> {
+    match store.delete_company(company.id.unwrap()).await {
+        Ok(_) => Ok(warp::reply::with_status(
+            "Delete Company Success".to_string(),
```
Comment (zk-steve) [inline src/controllers/company.rs:163]: .

File: src/controllers/job.rs [hunk 7]
```diff
@@ -0,0 +1,220 @@
+use crate::models::job::{Job, JobId, NewJob};
+use crate::models::map_resume_job::NewMapResumeJob;
+use crate::models::pagination::{Pagination, PaginationMethods};
+use crate::models::store_trait::StoreMethods;
+use crate::service::jwt::Claims;
+use crate::utils::convert_to_json::{Data, PayloadNoData, PayloadWithData};
+use reqwest::StatusCode;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tracing::instrument;
```
Comment (zk-steve) [inline src/controllers/job.rs:10]: .

File: src/controllers/job.rs [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+    }
+    let new_job = NewJob {
+        job_name: new_job.job_name,
+        company_id: new_job.company_id,
+        location: new_job.location,
+        quantity: new_job.quantity,
+        salary: new_job.salary,
+        job_level: new_job.job_level,
+        description: new_job.description,
+    };
```
Comment (zk-steve) [inline src/controllers/job.rs:42]: .

File: src/controllers/resume.rs [hunk 9]
```diff
@@ -0,0 +1,200 @@
+use crate::models::job::JobId;
+use crate::models::pagination::{Pagination, PaginationForJob, PaginationMethods};
+use crate::models::resume::{NewResume, Resume, ResumeId};
+use crate::models::store_trait::StoreMethods;
+use crate::service::jwt::Claims;
+use crate::utils::convert_to_json::{Data, PayloadNoData, PayloadWithData};
+use std::collections::HashMap;
+use std::sync::Arc;
+use tracing::instrument;
+use warp::http::StatusCode;
```
Comment (zk-steve) [inline src/controllers/resume.rs:10]: .

File: src/controllers/resume.rs [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+pub async fn create_resume(
+    store: Arc<dyn StoreMethods + Send + Sync>,
+    claims: Claims,
+    new_resume: NewResume,
+) -> Result<impl warp::Reply, warp::Rejection> {
+    let resume = NewResume {
+        user_id: new_resume.user_id,
+        email: new_resume.email,
+        url: new_resume.url,
+    };
```
Comment (zk-steve) [inline src/controllers/resume.rs:27]: .

File: src/controllers/user.rs [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+                                message: "Success".to_string(),
+                                data: Data::UserInfo(user_info),
+                            };
+                            Ok(warp::reply::with_status(
+                                warp::reply::json(&payload),
+                                StatusCode::OK,
+                            ))
+                        }
+                        Err(e) => Err(warp::reject::custom(e)),
+                    }
```
Comment (zk-steve) [inline src/controllers/user.rs:106]: optimize

File: src/main.rs [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+#[tokio::main]
+#[instrument]
+async fn main() {
+    let config = Config::new().expect("Config env not set");
+    let log_filter = format!(
+        "handle_errors={},rust-api-service={},warp={}",
+        config.log_level, config.log_level, config.log_level
+    );
+    init_telemetry(
+        config.service_name.as_str(),
```
Comment (zk-steve) [inline src/main.rs:34]: &
```suggestion
        &config.service_name,
```

File: src/main.rs [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        config.server.host.as_str(),
+        &config.server.jaeger_port,
+        log_filter.as_str(),
+    );
+
+    let store = build_store(&config).await;
+    let routes = build_routes(store).await.recover(return_error);
```
Comment (zk-steve) [inline src/main.rs:41]: remove async build_routes

File: src/main.rs [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+    });
+    tokio::task::spawn(server);
+    tx
+}
+
+pub async fn build_store_for_test(config: &Config) -> Arc<dyn StoreMethods + Send + Sync> {
+    let url = config.postgres.url.clone();
+
+    let store: Arc<dyn StoreMethods + Send + Sync> =
+        if config.database.clone().unwrap() == *"in-memory".to_string() {
```
Comment (zk-steve) [inline src/main.rs:112]:
```suggestion
        if &config.database == *"in-memory" {
```

File: src/main.rs [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+            info!("Using in-memory database");
+            Arc::new(InMemoryStore::new())
+        } else if config.database.clone().unwrap() == *"postgres".to_string() {
```
Comment (zk-steve) [inline src/main.rs:115]: else

File: src/models/store_in_memory.rs [hunk 16]
```diff
@@ -0,0 +1,547 @@
...
+        let id: i32 = if len == 0 { 1 } else { len + 1 };
+        let map_resume_job = MapResumeJob {
+            id: Some(MapResumeJobId(id)),
+            resume_id: new_map_resume_job.resume_id,
+            job_id: new_map_resume_job.job_id,
+        };
+        self.map_resume_job
+            .write()
+            .await
+            .insert(map_resume_job.id.clone().unwrap(), map_resume_job.clone());
```
Comment (zk-steve) [inline src/models/store_in_memory.rs:61]: data racing problem

File: src/routes/company.rs [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+    //Add base path into path
+    let company_path = warp::path(base_path)
+        .and(warp::path("v1"))
+        .and(warp::path("company"));
+    //Configures store filter
+    let store_filter = warp::any().map(move || store.clone());
+
+    //POST api/v1/company/createCompany
+    let create_api = company_path
+        .and(warp::path("createCompany"))
```
Comment (zk-steve) [inline src/routes/company.rs:24]: createCompany ?

File: src/service/handle_errors.rs [hunk 18]
```diff
@@ -0,0 +1,157 @@
+use tracing::{event, Level};
+use warp::{
+    filters::{body::BodyDeserializeError, cors::CorsForbidden},
+    http::StatusCode,
+    reject::Reject,
+    Rejection, Reply,
+};
+
+#[derive(Debug)]
+pub enum Error {
```
Comment (zk-steve) [inline src/service/handle_errors.rs:10]: anyhow, this error

File: src/service/handle_errors.rs [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+
+impl Reject for Error {}
+impl Reject for APILayerError {}
+
+// search in
+//https://www.ibm.com/docs/en/db2-for-zos/13?topic=codes-sqlstate-values-common-error#db2z_sqlstatevalues__classcode02
+const DUPLICATE_KEY: u32 = 23505;
+
+pub async fn return_error(r: Rejection) -> Result<impl Reply, Rejection> {
+    if let Some(Error::DatabaseQuery(e)) = r.find() {
```
Comment (zk-steve) [inline src/service/handle_errors.rs:83]: match

Comment (zk-steve) [inline src/service/handle_errors.rs:83]: r.find()?

File: src/tests/route_test.rs [hunk 20]
```diff
@@ -0,0 +1,934 @@
...
+use crate::models::user::{AuthInfo, UserId, UserInfo};
+use crate::utils::convert_to_json::PayloadForLogin;
+use crate::{build_store_for_test, init_mock_server};
+use futures_util::FutureExt;
+use tracing_subscriber::fmt::format::FmtSpan;
+
+#[tokio::test]
+async fn route_test() {
+    let config = Config::new().expect("Config env not set");
+    let address_listen = format!("{}:{}", config.server.host, config.server.port);
```
Comment (zk-steve) [inline src/tests/route_test.rs:15]: ...

File: src/main.rs [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+        &config.server.jaeger_port,
+        log_filter.as_str(),
+    );
+
+    let store = build_store(&config).await;
+    let routes = build_routes(store).await.recover(return_error);
+
+    let address_listen = format!("{}:{}", config.server.host, config.server.port);
+    let socket: std::net::SocketAddr = address_listen.parse().expect("Not a valid address");
+    warp::serve(routes).run(socket).await;
```
Comment (zk-steve) [inline src/main.rs:45]: rx.await.ok

File: src/tests/route_test.rs [hunk 22]
```diff
@@ -0,0 +1,934 @@
...
+    match std::panic::AssertUnwindSafe(get_resume_test(&access_token_resume))
+        .catch_unwind()
+        .await
+    {
+        Ok(_) => println!("✓"),
+        Err(_) => {
+            let _ = handler.send(1);
+            std::process::exit(1);
+        }
+    };
```
Comment (zk-steve) [inline src/tests/route_test.rs:432]: check returned data

File: .github/workflows/rust-ci.yml (hunk 1 above)
Comment (andrew2003) [reply .github/workflows/rust-ci.yml:23]: Done

File: Cargo.toml (hunk 2 above)
Comment (andrew2003) [reply Cargo.toml:33]: fmt cargo.toml done

File: src/controllers/company.rs (hunk 4 above)
Comment (andrew2003) [reply src/controllers/company.rs:39]: done

File: src/controllers/user.rs (hunk 11 above)
Comment (andrew2003) [reply src/controllers/user.rs:106]: Done

File: src/main.rs (hunk 13 above)
Comment (andrew2003) [reply src/main.rs:41]: Done

File: src/tests/route_test.rs (hunk 22 above)
Comment (andrew2003) [reply src/tests/route_test.rs:432]: done

File: src/routes/company.rs (hunk 17 above)
Comment (andrew2003) [reply src/routes/company.rs:24]: fix done

File: src/main.rs (hunk 15 above)
Comment (andrew2003) [reply src/main.rs:115]: done

**Pull request closed without merging**